File: ios/FastMqttFiletransferModule.swift
import ExpoModulesCore
import CocoaMQTT
import Foundation

public class FastMqttFiletransferModule: Module {
  private var mqttClient: CocoaMQTT5?
  private var socket: CocoaMQTTWebSocket?

  public func definition() -> ModuleDefinition {
    // 1. Module Name
    Name("FastMqttFiletransferModule")

    // 2. Declare events
    Events("onMqttStateChanged", "onFileTransferProgress")

    // 3. "initializeMqtt" function
    Function("initializeMqtt") { (scheme: String, host: String, port: UInt16, path: String?) -> Bool in
      // Validate scheme
      guard ["mqtt", "mqtts", "ws", "wss"].contains(scheme) else {
        throw FastMqttFiletransferModuleError.invalidScheme
      }

      let clientID = "ExpoMqttClient-\(UUID().uuidString)"

      // If using WebSockets
      if scheme == "ws" || scheme == "wss" {
        let wsURI = path ?? "/mqtt"
        let websocket = CocoaMQTTWebSocket(uri: wsURI)
        self.socket = websocket
        self.mqttClient = CocoaMQTT5(clientID: clientID, host: host, port: port, socket: websocket)
      } else {
        // Normal TCP
        self.mqttClient = CocoaMQTT5(clientID: clientID, host: host, port: port)
      }

      guard let mqttClient = self.mqttClient else {
        return false
      }

      // Enable SSL if needed
      mqttClient.enableSSL = (scheme == "mqtts" || scheme == "wss")
      mqttClient.allowUntrustCACertificate = true

      // Called when we get ConnAck from the broker
      mqttClient.didConnectAck = { [weak self] _, ack, _ in
        if ack == .success {
          self?.sendEvent("onMqttStateChanged", ["connected": true])
        } else {
          self?.sendEvent("onMqttStateChanged", ["connected": false])
        }
      }

      // Called when the client disconnects
      mqttClient.didDisconnect = { [weak self] _, _ in
        self?.sendEvent("onMqttStateChanged", ["connected": false])
      }

      // Attempt connecting
      return mqttClient.connect()
    }

    // 4. "sendFile" async function
    AsyncFunction("sendFile") {
      (filePath: String, encoding: String, destinationTopic: String, chunkSize: Int?, chunkIndex: Int?) -> Bool in

      guard let mqttClient = self.mqttClient else {
        throw FastMqttFiletransferModuleError.notConnected
      }
      guard mqttClient.connState == .connected else {
        throw FastMqttFiletransferModuleError.notConnected
      }

      // Validate file
      let url = URL(fileURLWithPath: filePath)
      guard FileManager.default.fileExists(atPath: url.path) else {
        throw FastMqttFiletransferModuleError.fileNotFound
      }
      let fileData = try Data(contentsOf: url)

      // Either single chunk or native chunk
      if let chunkSize = chunkSize, let chunkIndex = chunkIndex {
        // Single-chunk
        let start = chunkIndex * chunkSize
        let end = min(start + chunkSize, fileData.count)
        guard start < fileData.count else {
          throw FastMqttFiletransferModuleError.invalidChunkRange
        }

        let chunkData = fileData.subdata(in: start ..< end)
        try self.publishData(chunkData, encoding: encoding, topic: destinationTopic)
        return true

      } else {
        // Native chunking
        let chunk = chunkSize ?? fileData.count
        let totalChunks = Int(ceil(Double(fileData.count) / Double(chunk)))
        var sentChunks = 0

        for start in stride(from: 0, to: fileData.count, by: chunk) {
          let end = min(start + chunk, fileData.count)
          let chunkData = fileData.subdata(in: start ..< end)

          try self.publishData(chunkData, encoding: encoding, topic: destinationTopic)

          sentChunks += 1
          let percentage = (Double(sentChunks) / Double(totalChunks)) * 100
          self.sendEvent("onFileTransferProgress", [
            "chunkIndex": sentChunks - 1,
            "totalChunks": totalChunks,
            "percentage": percentage
          ])
        }
        return true
      }
    }

    // 5. A function to publish a test message
    Function("publishTestMessage") { (topic: String) -> Bool in
      guard let mqttClient = self.mqttClient, mqttClient.connState == .connected else {
        throw FastMqttFiletransferModuleError.notConnected
      }
      // Build a test CocoaMQTT5Message
      let message = CocoaMQTT5Message(topic: topic, string: "Hello from Swift!")
      let props = MqttPublishProperties()
      _ = mqttClient.publish(message, DUP: false, retained: false, properties: props)
      return true
    }
  }

  // MARK: - Helpers
  private func publishData(_ data: Data, encoding: String, topic: String) throws {
    guard let mqttClient = mqttClient else {
      throw FastMqttFiletransferModuleError.notConnected
    }

    // Determine final payload (Base64, UTF8, or fallback to raw)
    let payload: String
    switch encoding.lowercased() {
      case "base64":
        payload = data.base64EncodedString()
      case "utf8":
        guard let text = String(data: data, encoding: .utf8) else {
          throw FastMqttFiletransferModuleError.encodingFailed
        }
        payload = text
      default:
        payload = String(decoding: data, as: UTF8.self)
    }

    // Publish using CocoaMQTT5
    let message = CocoaMQTT5Message(topic: topic, string: payload)
    let props = MqttPublishProperties()
    _ = mqttClient.publish(message, DUP: false, retained: false, properties: props)
  }
}

// MARK: - Errors
enum FastMqttFiletransferModuleError: Error, CustomStringConvertible {
  case notConnected
  case fileNotFound
  case invalidChunkRange
  case encodingFailed
  case invalidScheme

  var description: String {
    switch self {
    case .notConnected:
      return "MQTT client is not connected."
    case .fileNotFound:
      return "File not found at the given path."
    case .invalidChunkRange:
      return "The requested chunk is out of range."
    case .encodingFailed:
      return "Failed to apply the requested encoding."
    case .invalidScheme:
      return "Invalid scheme. Supported schemes are 'mqtt', 'mqtts', 'ws', 'wss'."
    }
  }
}

File: android/src/main/java/expo/modules/fastmqttfiletransfer/FastMqttFiletransferModule.kt
package expo.modules.fastmqttfiletransfer

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.os.Build
import android.util.Log
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import org.eclipse.paho.android.service.MqttAndroidClient
import org.eclipse.paho.client.mqttv3.*
import java.io.File
import java.util.*

class FastMqttFiletransferModule : Module() {

  companion object {
    const val MQTT_STATE_CHANGED_EVENT = "onMqttStateChanged"
    const val FILE_TRANSFER_PROGRESS_EVENT = "onFileTransferProgress"
  }

  private var mqttClient: MqttAndroidClient? = null
  private var isConnected = false // Change val to var to allow reassignment

  /**
   * Helper function to encode data according to 'encoding'.
   */
  private fun encodeData(data: ByteArray, encoding: String): ByteArray {
    return when (encoding.lowercase()) {
      "base64" -> android.util.Base64.encode(data, android.util.Base64.NO_WRAP)
      "utf8" -> {
        val text = String(data, Charsets.UTF_8)
        text.toByteArray(Charsets.UTF_8)
      }
      else -> data
    }
  }

  /**
   * Helper function to publish a chunk via MQTT.
   */
  private fun publishChunk(payload: ByteArray, client: MqttAndroidClient, topic: String) {
    val message = MqttMessage(payload).apply {
      qos = 1
      isRetained = false
    }
    client.publish(topic, message, null, object : IMqttActionListener {
      override fun onSuccess(asyncActionToken: IMqttToken?) {
        Log.d("FastMqttFiletransferModule", "Chunk published successfully.")
      }

      override fun onFailure(asyncActionToken: IMqttToken?, exception: Throwable?) {
        Log.e("FastMqttFiletransferModule", "Failed to publish chunk: ${exception?.message}")
      }
    })
  }

  /**
   * Helper function to send the MQTT connection state.
   */
  private fun sendMQTTStateEvent(connected: Boolean) {
    sendEvent(
      MQTT_STATE_CHANGED_EVENT,
      mapOf("connected" to connected)
    )
  }

  /**
   * Helper function to send the file transfer progress.
   */
  private fun sendFileTransferProgressEvent(chunkIndex: Int, totalChunks: Long) {
    val percent = ((chunkIndex + 1) * 100) / totalChunks.toDouble()
    sendEvent(
      FILE_TRANSFER_PROGRESS_EVENT,
      mapOf(
        "chunkIndex" to chunkIndex,
        "totalChunks" to totalChunks,
        "percentage" to percent
      )
    )
  }

  override fun definition() = ModuleDefinition {
    Name("FastMqttFiletransferModule")

    Events(MQTT_STATE_CHANGED_EVENT, FILE_TRANSFER_PROGRESS_EVENT)

    Function("initializeMqtt") { scheme: String, host: String, port: Int, path: String? ->
      // Validate scheme
      if (scheme !in listOf("mqtt", "mqtts", "ws", "wss")) {
        throw IllegalArgumentException("Invalid scheme: $scheme. Use 'mqtt', 'mqtts', 'ws', or 'wss'.")
      }

      val serverUri = if (scheme == "ws" || scheme == "wss") {
        "$scheme://$host:$port${path ?: "/mqtt"}"
      } else {
        "$scheme://$host:$port"
      }

      val clientId = "ExpoMqttClient-" + UUID.randomUUID().toString()
      val context = requireNotNull(appContext.reactContext) {
        "React context is null."
      }

      mqttClient = MqttAndroidClient(context, serverUri, clientId).apply {
        setCallback(object : MqttCallbackExtended {
          override fun connectComplete(reconnect: Boolean, serverURI: String?) {
            this@FastMqttFiletransferModule.isConnected = true
            Log.d("FastMqttModule", "connectComplete, serverURI=$serverURI")
            sendMQTTStateEvent(true)
          }

          override fun connectionLost(cause: Throwable?) {
            this@FastMqttFiletransferModule.isConnected = false
            Log.e("FastMqttModule", "connectionLost: ${cause?.message}")
            sendMQTTStateEvent(false)
          }

          override fun messageArrived(topic: String, message: MqttMessage) {}
          override fun deliveryComplete(token: IMqttDeliveryToken) {}
        })
      }

      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
        val channelId = "mqtt_channel"
        val channel = NotificationChannel(
          channelId, "MQTT Notifications", NotificationManager.IMPORTANCE_LOW
        )
        val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        notificationManager.createNotificationChannel(channel)

        val notification = Notification.Builder(context, channelId)
          .setContentTitle("MQTT Service")
          .setContentText("Running")
          .setSmallIcon(android.R.drawable.ic_menu_info_details)
          .build()

        mqttClient?.setForegroundService(notification, 1)
      }

      val options = MqttConnectOptions().apply {
        isAutomaticReconnect = true
        isCleanSession = true
      }

      try {
        mqttClient?.connect(options, null, object : IMqttActionListener {
          override fun onSuccess(asyncActionToken: IMqttToken?) {
            this@FastMqttFiletransferModule.isConnected = true
            Log.d("FastMqttModule", "Connected to MQTT broker")
            sendMQTTStateEvent(true)
          }

          override fun onFailure(asyncActionToken: IMqttToken?, exception: Throwable?) {
            this@FastMqttFiletransferModule.isConnected = false
            Log.e("FastMqttModule", "Failed to connect: ${exception?.message}")
            sendMQTTStateEvent(false)
          }
        })
      } catch (e: Exception) {
        Log.e("FastMqttModule", "initializeMqtt error: $e")
      }
      true
    }

    // 2. Send File (supports single-chunk or multi-chunk)
    AsyncFunction("sendFile") {
        filePath: String,
        encoding: String,
        destinationTopic: String,
        chunkSize: Int,
        chunkIndex: Int? ->

      // Basic checks
      if (!isConnected) {
        throw Exception("MQTT client is not connected.")
      }
      val client = mqttClient ?: throw Exception("MQTT client is not initialized.")

      // Validate file and chunk size
      val file = File(filePath)
      if (!file.exists()) {
        throw Exception("File not found at $filePath")
      }
      if (chunkSize <= 0) {
        throw IllegalArgumentException("Chunk size must be > 0")
      }

      // SINGLE-CHUNK Mode
      if (chunkIndex != null) {
        val fileSize = file.length() // Long
        val startLong = chunkIndex.toLong() * chunkSize.toLong()
        if (startLong >= fileSize) {
          throw Exception("Requested chunk is out of range.")
        }

        val endLong = minOf(startLong + chunkSize.toLong(), fileSize)
        val bytesToRead = (endLong - startLong).toInt()

        // Read just that chunk
        val data = file.inputStream().use { input ->
          input.skip(startLong)
          input.readNBytes(bytesToRead)
        }

        val payload = encodeData(data, encoding)
        publishChunk(payload, client, destinationTopic)

        true

      } else {
        // MULTI-CHUNK STREAMING Mode

        val fileSize = file.length()
        val totalChunks = (fileSize + chunkSize - 1) / chunkSize

        file.inputStream().buffered().use { inputStream ->
          val buffer = ByteArray(chunkSize)
          var bytesRead: Int
          var chunkIndex = 0

          while (inputStream.read(buffer).also { bytesRead = it } != -1) {
            val chunkData = buffer.copyOf(bytesRead)
            val payload = encodeData(chunkData, encoding)

            publishChunk(payload, client, destinationTopic)
            Log.d("FastMqttFiletransferModule", "Sent chunk $chunkIndex")

            sendFileTransferProgressEvent(chunkIndex, totalChunks)

            chunkIndex++
          }
        }
        true
      }
    }

    Function("publishTestMessage") { topic: String ->
      if (!isConnected) throw Exception("MQTT client is not connected.")
      val client = mqttClient ?: throw Exception("MQTT client is not initialized.")

      val message = MqttMessage("Hello from Kotlin!".toByteArray(Charsets.UTF_8)).apply {
        qos = 1
        isRetained = false
      }
      client.publish(topic, message)
      true
    }
  }
}